x/payment/types/wirepayfordata.go

```diff
@@ -118,9 +130,36 @@ func (msg *MsgWirePayForData) ValidateBasic() error {
 		}
 	}
 
+	if err := msg.ValidateAllSquareSizesCommitedTo(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// ValidateAllSquareSizesCommitedTo returns an error if the list of square sizes
+// committed to don't match all squares sizes expected for this message size.
+func (msg *MsgWirePayForData) ValidateAllSquareSizesCommitedTo() error {
+	allSquareSizes := AllSquareSizes(int(msg.MessageSize))
+	sort.Slice(allSquareSizes, func(i, j int) bool { return allSquareSizes[i] < allSquareSizes[j] })
+
+	committedSquareSizes := msg.committedSquareSizes()
+	sort.Slice(committedSquareSizes, func(i, j int) bool { return committedSquareSizes[i] < committedSquareSizes[j] })
```

Comment:
[for reviewers] sorting isn't strictly necessary here. We could convert these slices to sets (perhaps backed by a `map[uint64]bool`) and implement an equality operator for these sets. 

Performance would improve from O(N log N) for sorting to O(N) b/c constant time operations for all elements in `allSquareSizes` + `committedSquareSizes`. This type of optimization seems premature given the length of `allSquareSizes` is currently capped to 8. `commitedSquareSizes` is potentially unbounded so we should consider adding a check for that being too large 😳

Reply:
> is potentially unbounded so we should consider adding a check for that being too large 

Resolved in https://github.com/celestiaorg/celestia-app/pull/666/commits/baa170ba27f64164ec2b9cc5ef446723e3547312